:zap: use qs_dart for query string encoding

techouse opened the pull request:
@diegotori noticed that a `Map<String, dynamic>` like this
```dart
{
  'filters': {
    r'$or': [
      {
        'date': {
          r'$eq': '2020-01-01',
        }
      },
      {
        'date': {
          r'$eq': '2020-01-02',
        }
      }
    ],
    'author': {
      'name': {
        r'$eq': 'Kai doe',
      },
    }
  }
}
```
did not get encoded properly to
```
filters[$or][][date][$eq]=2020-01-01&filters[$or][][date][$eq]=2020-01-02&filters[author][name][$eq]=Kai%20doe
```
but became this
```
filters[$or][]={date: {$eq: 2020-01-01}}&filters[$or][]={date: {$eq: 2020-01-02}}&filters[author][name][$eq]=Kai doe
```

This PR fixes this by using an external library [qs_dart](https://pub.dev/packages/qs_dart) which is a Dart port of [qs](https://www.npmjs.com/package/qs) which I've been painstakingly working on over the past couple of weeks. 🤓

This now means that the output of Chopper, while still backwards compliant, is now [qs](https://www.npmjs.com/package/qs) compatible.

The only **caveat** is `null` and empty `String` handling in query strings, which is a bit different in [qs](https://www.npmjs.com/package/qs) and therefore also [qs_dart](https://pub.dev/packages/qs_dart).

Please check the various test case changes to see these `null` and empty `String` differences.

---

Addresses #584
Supersedes #591

CC / @JEuler @Guldem @diegotori

techouse commented:
I propose this goes into v8.0.0 since it slightly changes the way `null` and empty `String`s are encoded.

CC / @JEuler @Guldem @diegotori

Guldem commented:
> I propose this goes into v8.0.0 since it slightly changes the way `null` and empty `String`s are encoded.
> 
> CC / @JEuler @Guldem @diegotori

Makes sense. Introduce all changes at once 😉 

LGTM! Also want someone to have a look at [qs_dart](https://pub.dev/packages/qs_dart)?

techouse commented:
> LGTM! Also want someone to have a look at [qs_dart](https://pub.dev/packages/qs_dart)?

Please :) I converted [all the tests](https://github.com/ljharb/qs/tree/v6.12.0/test) of [qs](https://www.npmjs.com/package/qs), however, porting the JavaScript code was quite the ordeal since it's not typed and written in such a legacy way that made me cry a few times 🙈

diegotori commented:
@techouse I still need to add the encoding tests that evaluate Strapi query map encoding into this library's tests.

Also, please consider adding an option for indexed bracketed arrays, unless `qs_dart` already does that by default, then we're already good.

techouse commented:
> Also, please consider adding an option for indexed bracketed arrays, unless `qs_dart` already does that by default, then we're already good.

`qs_dart` indeed defaults to `ListFormat.indices`, however, Chopper does not.

I'm working on deprecating the current tags (while still keeping them) and adding a new ones which will give the user an option to encode with any of [these](https://pub.dev/documentation/qs_dart/latest/qs_dart/ListFormat.html):
- brackets (current default)
- repeat
- indices
- comma

diegotori commented:
@techouse I think we should also target this for `7.4.0`, so that it's available immediately.

Otherwise, the changes related to the interceptors might get in the way if we have to upgrade to `8.0.0`.

diegotori commented:
Add the following test groups to `utils_test.dart`:

```dart

  group('mapToQuery maps with brackets and nested lists', () {
    <Map<String, dynamic>, String>{
      {
        'filters': {
          r'$or': [
            {
              'date': {
                r'$eq': '2020-01-01',
              }
            },
            {
              'date': {
                r'$eq': '2020-01-02',
              }
            }
          ],
          'author': {
            'name': {
              r'$eq': 'Kai doe',
            },
          }
        }
      }: 'filters%5B%24or%5D%5B%5D%5Bdate%5D%5B%24eq%5D=2020-01-01&filters%5B%24or%5D%5B%5D%5Bdate%5D%5B%24eq%5D=2020-01-02&filters%5Bauthor%5D%5Bname%5D%5B%24eq%5D=Kai%20doe',
      {
        'filters': {
          'id': {
            r'$in': [3, 6, 8],
          },
        }
      }: 'filters%5Bid%5D%5B%24in%5D%5B%5D=3&filters%5Bid%5D%5B%24in%5D%5B%5D=6&filters%5Bid%5D%5B%24in%5D%5B%5D=8'
    }.forEach(
      (map, query) => test(
        '$map -> $query',
        () => expect(
          mapToQuery(map, useBrackets: true),
          query,
        ),
      ),
    );
  });

  group(
      'mapToQuery maps with brackets with includeNullQueryVars and nested lists',
      () {
    <Map<String, dynamic>, String>{
      {
        'filters': {
          r'$or': [
            {
              'date': {
                r'$eq': '2020-01-01',
              }
            },
            null,
            {
              'date': {
                r'$eq': '2020-01-02',
              }
            }
          ],
          'author': {
            'name': {
              r'$eq': 'Kai doe',
            },
          }
        }
      }: 'filters%5B%24or%5D%5B%5D%5Bdate%5D%5B%24eq%5D=2020-01-01&filters%5B%24or%5D%5B%5D=&filters%5B%24or%5D%5B%5D%5Bdate%5D%5B%24eq%5D=2020-01-02&filters%5Bauthor%5D%5Bname%5D%5B%24eq%5D=Kai%20doe',
      {
        'filters': {
          'id': {
            r'$in': [3, null, 8],
          },
        }
      }: 'filters%5Bid%5D%5B%24in%5D%5B%5D=3&filters%5Bid%5D%5B%24in%5D%5B%5D=&filters%5Bid%5D%5B%24in%5D%5B%5D=8'
    }.forEach(
      (map, query) {
        test(
          '$map -> $query',
          () => expect(
            mapToQuery(map, useBrackets: true, includeNullQueryVars: true),
            query,
          ),
        );
      },
    );
  });
```

techouse commented:
> @techouse I think we should also target this for `7.4.0`, so that it's available immediately.
> 
> Otherwise, the changes related to the interceptors might get in the way if we have to upgrade to `8.0.0`.
> 
> Not to mention, it gives you time to deprecate `useBrackets` and replace it with `ListFormat`.

Hmm, fair point.

@Guldem are you onboard with this?

techouse commented on chopper/test/utils_test.dart:
```diff
@@ -554,6 +617,27 @@ void main() {
+          }
+        }
+      }: 'filters%5B%24or%5D%5B%5D%5Bdate%5D%5B%24eq%5D=2020-01-01&filters%5B%24or%5D%5B%5D%5Bdate%5D%5B%24eq%5D=2020-01-02&filters%5Bauthor%5D%5Bname%5D%5B%24eq%5D=John%20doe',
     }.forEach(
```
Done.

Guldem commented:
> > @techouse I think we should also target this for `7.4.0`, so that it's available immediately.
> > Otherwise, the changes related to the interceptors might get in the way if we have to upgrade to `8.0.0`.
> > Not to mention, it gives you time to deprecate `useBrackets` and replace it with `ListFormat`.
> 
> Hmm, fair point.
> 
> @Guldem are you onboard with this?

That's fine! This is needed, the interceptors not necessarily 😄

diegotori commented on chopper/lib/src/utils.dart:
```diff
@@ -66,98 +66,19 @@ String mapToQuery(
-          ? value.toUtc().toIso8601String()
-          : value?.toString() ?? '',
+      EncodeOptions(
+        listFormat: useBrackets ? ListFormat.brackets : ListFormat.repeat,
```
If you want to deprecate `useBrackets`, consider adding the following:

```dart
String mapToQuery(
  Map<String, dynamic> map, {
  @Deprecated("Use listFormat instead. Will be removed in a future release.")
  bool useBrackets = false,
  bool includeNullQueryVars = false,
  ListFormat listFormat = ListFormat.repeat,
}) =>
    QS.encode(
      map,
      EncodeOptions(
        listFormat: useBrackets ? ListFormat.brackets : listFormat,
        allowDots: !useBrackets,
        encodeDotInKeys: !useBrackets,
        encodeValuesOnly: !useBrackets,
        skipNulls: !includeNullQueryVars,
        strictNullHandling: false,
        serializeDate: (DateTime date) => date.toUtc().toIso8601String(),
      ),
);
```

techouse replied:
Way ahead of you buddy :)

techouse commented:
I've added a new annotation `listFormat`
```dart
/// List format to use when encoding lists
///
/// - [ListFormat.repeat] (default) hxxp://path/to/script?foo=123&foo=456&foo=789
/// - [ListFormat.brackets] hxxp://path/to/script?foo[]=123&foo[]=456&foo[]=789
/// - [ListFormat.indices] hxxp://path/to/script?foo[0]=123&foo[1]=456&foo[2]=789
/// - [ListFormat.comma] hxxp://path/to/script?foo=123,456,789
final ListFormat? listFormat;
```
and have deprecated `useBrackets`
```dart
/// Use brackets [ ] to when encoding
///
/// - lists
/// hxxp://path/to/script?foo[]=123&foo[]=456&foo[]=789
///
/// - maps
/// hxxp://path/to/script?user[name]=john&user[surname]=doe&user[age]=21
@Deprecated('Use listFormat instead')
final bool? useBrackets;
```
In line with the above I have also modified `mapToQuery`
```dart
String mapToQuery(
  Map<String, dynamic> map, {
  ListFormat? listFormat,
  @Deprecated('Use listFormat instead') bool? useBrackets,
  bool? includeNullQueryVars,
}) {
  listFormat ??= useBrackets == true ? ListFormat.brackets : ListFormat.repeat;

  return QS.encode(
    map,
    EncodeOptions(
      listFormat: listFormat,
      allowDots: listFormat == ListFormat.repeat,
      encodeDotInKeys: listFormat == ListFormat.repeat,
      encodeValuesOnly: listFormat == ListFormat.repeat,
      skipNulls: includeNullQueryVars != true,
      strictNullHandling: false,
      serializeDate: (DateTime date) => date.toUtc().toIso8601String(),
    ),
  );
}
```
and added a tonne of tests 🤪

CC / @JEuler @Guldem @diegotori

techouse commented:
@diegotori could you please test this PR in your problematic project/app and report back any findings?
```yaml
dependency_overrides:
  chopper:
    git:
      url: https://github.com/Guldem/chopper.git
      ref: fix/issue-584-as
      path: chopper
  chopper_generator:
    git:
      url: https://github.com/Guldem/chopper.git
      ref: fix/issue-584-as
      path: chopper_generator
```

techouse commented:
I think I'll have to somehow export Chopper's own `ListFormat` instead of using the one from `qs_dart` directly.

diegotori commented:
> ```yaml
> dependency_overrides:
>   chopper:
>     git:
>       url: https://github.com/lejard-h/chopper.git
>       ref: fix/issue-584-qs
>       path: chopper
>   chopper_generator:
>     git:
>       url: https://github.com/lejard-h/chopper.git
>       ref: fix/issue-584-qs
>       path: chopper_generator
> ```

I dunno if I have the capacity to do that currently. However, once you release it and we start the process of migrating to using this new version, if there are any issues, we'll be sure to raise them.

So far, the tests that were added to both this library and `qs_dart` are reflective of the issues that we ran into when trying to create query strings from maps in chopper.

Since `qs_dart` is literally a 1:1 port of JavaScript's `qs` library (which Strapi uses), then I'm highly confident that this new version will work flawlessly.

techouse commented:
> Since qs_dart is literally a 1:1 port of JavaScript's qs library (which Strapi uses), then I'm highly confident that this new version will work flawlessly.

I tried my best to port everything that could be ported, however, I could not make it a 1:1 copy because some core differences between JavaScript and Dart, i.e. [Object prototypes](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Object_prototypes), [sparse arrays](https://medium.com/@shore.jeremiah/what-the-hell-is-a-sparse-array-in-javascript-and-why-might-it-matter-44f950119f27), etc.

diegotori commented:
> > Since qs_dart is literally a 1:1 port of JavaScript's qs library (which Strapi uses), then I'm highly confident that this new version will work flawlessly.
> 
> I tried my best to port everything that could be ported, however, I could not make it a 1:1 port because of some core differences between JavaScript and Dart, i.e. [Object prototypes](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Object_prototypes), [sparse arrays](https://medium.com/@shore.jeremiah/what-the-hell-is-a-sparse-array-in-javascript-and-why-might-it-matter-44f950119f27), etc.

No worries. So far, it's spitting out Strapi-compatible query strings, and that's what really matters for us.

That way, we won't have to do backend hackery to get around the previous limitations.

Standby for another set of tests that you'll need to add so that it covers ones that I found.

diegotori commented on chopper/test/utils_test.dart:
```diff
@@ -554,17 +933,702 @@ void main() {
+    );
+  });
+
+  group('mapToQuery maps with indices and nested lists', () {
```
Add the following tests in order to cover the new `ListFormat` based formatting when encoding nested lists:

```dart

  group(
    'mapToQuery maps with repeat (default) and nested lists',
    () {
      <Map<String, dynamic>, String>{
        {
          'filters': {
            r'$or': [
              {
                'date': {
                  r'$eq': '2020-01-01',
                }
              },
              null,
              {
                'date': {
                  r'$eq': '2020-01-02',
                }
              }
            ],
            'author': {
              'name': {
                r'$eq': 'Kai doe',
              },
            }
          }
        }: r'filters%2E$or%2Edate.$eq=2020-01-01&filters%2E$or%2Edate.$eq=2020-01-02&filters%2Eauthor%2Ename.$eq=Kai%20doe',
        {
          'filters': {
            'id': {
              r'$in': [3, 6, 8],
            },
          }
        }: r'filters%2Eid%2E$in=3&filters%2Eid%2E$in=6&filters%2Eid%2E$in=8'
      }.forEach(
        (map, query) {
          test(
            '$map -> $query',
            () => expect(
              mapToQuery(
                map,
              ),
              query,
              reason: 'legacy default',
            ),
          );

          test(
            '$map -> $query',
            () => expect(
              mapToQuery(
                map,
                listFormat: ListFormat.repeat,
              ),
              query,
            ),
          );
        },
      );
    },
  );

  group(
    'mapToQuery maps with brackets and nested lists',
    () {
      <Map<String, dynamic>, String>{
        {
          'filters': {
            r'$or': [
              {
                'date': {
                  r'$eq': '2020-01-01',
                }
              },
              {
                'date': {
                  r'$eq': '2020-01-02',
                }
              }
            ],
            'author': {
              'name': {
                r'$eq': 'Kai doe',
              },
            }
          }
        }: 'filters%5B%24or%5D%5B%5D%5Bdate%5D%5B%24eq%5D=2020-01-01&filters%5B%24or%5D%5B%5D%5Bdate%5D%5B%24eq%5D=2020-01-02&filters%5Bauthor%5D%5Bname%5D%5B%24eq%5D=Kai%20doe',
        {
          'filters': {
            'id': {
              r'$in': [3, 6, 8],
            },
          }
        }: 'filters%5Bid%5D%5B%24in%5D%5B%5D=3&filters%5Bid%5D%5B%24in%5D%5B%5D=6&filters%5Bid%5D%5B%24in%5D%5B%5D=8'
      }.forEach(
        (map, query) {
          test(
            '$map -> $query',
            () => expect(
              mapToQuery(
                map,
                useBrackets: true,
              ),
              query,
              reason: 'legacy brackets',
            ),
          );

          test(
            '$map -> $query',
            () => expect(
              mapToQuery(
                map,
                listFormat: ListFormat.brackets,
              ),
              query,
            ),
          );
        },
      );
    },
  );

  group(
    'mapToQuery maps with comma and nested lists',
    () {
      <Map<String, dynamic>, String>{
        {
          'filters': {
            r'$or': [
              {
                'date': {
                  r'$eq': '2020-01-01',
                }
              },
              {
                'date': {
                  r'$eq': '2020-01-02',
                }
              }
            ],
            'author': {
              'name': {
                r'$eq': 'Kai doe',
              },
            }
          }
        }: 'filters%5B%24or%5D=%7Bdate%3A%20%7B%24eq%3A%202020-01-01%7D%7D%2C%7Bdate%3A%20%7B%24eq%3A%202020-01-02%7D%7D&filters%5Bauthor%5D%5Bname%5D%5B%24eq%5D=Kai%20doe',
        {
          'filters': {
            'id': {
              r'$in': [3, 6, 8],
            },
          }
        }: 'filters%5Bid%5D%5B%24in%5D=3%2C6%2C8'
      }.forEach(
        (map, query) => test(
          '$map -> $query',
          () => expect(
            mapToQuery(
              map,
              listFormat: ListFormat.comma,
              includeNullQueryVars: true,
            ),
            query,
          ),
        ),
      );
    },
  );
```

techouse replied:
Done.

diegotori reviewed (approved):
LGTM.

techouse commented:
I identified another flaw. I forgot to serialize any object to string.
